monai/apps/auto3dseg/auto_runner.py

```diff
@@ -181,58 +181,80 @@ def __init__(
+
+        logger.info(f"AutoRunner using work directory {work_dir}")
+        os.makedirs(work_dir, exist_ok=True)
 
         self.work_dir = os.path.abspath(work_dir)
         self.data_src_cfg_name = os.path.join(self.work_dir, "input.yaml")
+        self.algos = algos
 
-        if input is None:
-            input = os.path.join(self.work_dir, "input.yaml")
-        elif isinstance(input, Dict):
-            ConfigParser.export_config_file(input, self.data_src_cfg_name)
+        if input is None and os.path.isfile(self.data_src_cfg_name):
+            input = self.data_src_cfg_name
+            logger.info(f"Input config is not provided, using the default {input}")
+
+        if isinstance(input, Dict):
             self.data_src_cfg = input
+            ConfigParser.export_config_file(
+                config=input, filepath=self.data_src_cfg_name, fmt="yaml", default_flow_style=None, sort_keys=False
+            )
         elif isinstance(input, str) and os.path.isfile(input):
-            shutil.copy(input, self.data_src_cfg_name)
-            logger.info(f"Loading {input} for AutoRunner and making a copy in {self.data_src_cfg_name}")
-            self.data_src_cfg = ConfigParser.load_config_file(self.data_src_cfg_name)
+            self.data_src_cfg = ConfigParser.load_config_file(input)
+            logger.info(f"Loading input config {input}")
+            if input != self.data_src_cfg_name:
+                shutil.copy(input, self.data_src_cfg_name)
         else:
-            raise ValueError(f"{input} is not a valid file")
+            raise ValueError(f"{input} is not a valid file or dict")
+
+        missing_keys = {"dataroot", "datalist", "modality"}.difference(self.data_src_cfg.keys())
+        if len(missing_keys) > 0:
+            raise ValueError(f"Config keys are missing {missing_keys}")
+
+        self.dataroot = self.data_src_cfg["dataroot"]
+        self.datalist_filename = self.data_src_cfg["datalist"]
+        self.datastats_filename = os.path.join(self.work_dir, "datastats.yaml")
 
         self.not_use_cache = not_use_cache
         self.cache_filename = os.path.join(self.work_dir, "cache.yaml")
-        self.cache = self.check_cache()
+        self.cache = self.read_cache()
         self.export_cache()
 
-        # Whether we need all the steps or not
-        self.analyze = self.check_analyze(analyze)
-        self.algo_gen = self.check_algo_gen(algo_gen)
-        self.train = self.check_train(train)
+        # determine if we need to analyze, algo_gen or train from cache, unless manually provided
+        self.analyze = not self.cache["analyze"] if analyze is None else analyze
+        self.algo_gen = not self.cache["algo_gen"] if algo_gen is None else algo_gen
+        self.train = not self.cache["train"] if train is None else train
         self.ensemble = ensemble  # last step, no need to check
 
+        # since each next depends on the previous, check the dependencies
+        self.train = self.train or (
```

Comment:
Hi @myron , thanks for adding the checks here. I have a question here about the behavior of `AutoRunner` here.  If the user sets the `train` to `False` but `ensemble` to `True`, and `AutoRunner` would start re-training, which may lead to file overwriting. Is it better to ask, or simply fail with error messages?

Reply:
good question. right now it looks like this
```
  self.train = self.train or (
      self.ensemble and not self.cache["train"]
  )  
```
in your example, if the user sets the `train = False`, but `ensemble = True`,  than there is another condition : if cache indicates the training was completed,  the train ramains False, but if cache indicates the training was Not completed then train becomes True (but it's a necessary step for ensembling, and it won't overwrite anything since we haven't finished this step yet).   We can always shows an error, but it wont be a very smooth workflow. what do you think?

Reply:
I've removed these assignments, and now it'll raise an error if the previous step was missing